### PR TITLE
feat: implement bulkhead

### PR DESCRIPTION
**Pull request by connor4312**
Nothing surprising here. See the readme for usage details.

### CODE REVIEW

File: src/errors/BulkheadRejectedError.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+export class BulkheadRejectedError extends Error {
+  constructor() {
+    super('Bulkhead capacity exceeded');
```
Comment (SimonSchick) [inline src/errors/BulkheadRejectedError.ts]: Nit: Use `public readonly message = '...'`

File: src/Bulkhead.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+  resolve(value: T): void;
+  reject(error: Error): void;
+}
+
+/**
+ * Bulkhead limits concurrent requests made.
+ */
+export class Bulkhead {
+  private active = 0;
+  private queue: Array<IQueueItem<any>> = [];
```
Comment (SimonSchick) [inline src/Bulkhead.ts]: Probably `unknown`.

File: .vscode/launch.json [hunk 3]
```diff
@@ -5,14 +5,13 @@
   "version": "0.2.0",
```
Comment (SimonSchick) [inline .vscode/launch.json]: git diff warning here.

File: src/Bulkhead.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    if (this.queue.length < this.queueCapacity) {
+      const { resolve, reject, promise } = defer<T>();
+      this.queue.push({ fn, resolve, reject });
+      return promise;
+    }
+
+    this.onRejectEmitter.emit();
+    throw new BulkheadRejectedError();
```
Comment (SimonSchick) [inline src/Bulkhead.ts]: Maybe include the capacity in the error as a public readonly and emit that error through the emitter as well?

File: src/Policy.ts [hunk 5]
```diff
@@ -26,6 +27,13 @@ export type FailureReason<R> = { error: Error } | { value: R };
  * breakers, retries, etc.
  */
 export class Policy<ReturnType> {
+  /**
+   * Creates a bulkhead--a policy that limits the number of concurrent calls.
+   */
+  public static bulkhead(limit: number, queue: number = 0) {
```
Comment (SimonSchick) [inline src/Policy.ts]: So the bulkhead provides both throttling and limiting of parallelism?

File: .vscode/launch.json (hunk 3 above)
Comment (connor4312) [reply .vscode/launch.json]: it's allowed :P

File: src/Policy.ts (hunk 5 above)
Comment (connor4312) [reply src/Policy.ts]: Just limiting of parallelism, with an optional queue to avoid breaking immediately in bursty scenarios.

**Pull request merged**